server/src/main/java/org/opensearch/gateway/remote/routingtable/IndexRoutingTableInputStream.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway.remote.routingtable;
+
+import org.opensearch.Version;
+import org.opensearch.cluster.routing.IndexRoutingTable;
+import org.opensearch.cluster.routing.IndexShardRoutingTable;
+import org.opensearch.common.io.stream.BufferedChecksumStreamOutput;
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.core.common.bytes.BytesReference;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+
+public class IndexRoutingTableInputStream extends InputStream {
+
+    /**
+     * The buffer where data is stored.
+     */
+    protected byte[] buf;
+
+    /**
+     * The number of valid bytes in the buffer.
+     */
+    protected int count;
+
+    /**
+     * The buffer left over from the last fill
+     */
+    protected byte[] leftOverBuf;
+
+    /**
+     * The mark position
+     */
+    protected int markPos = -1;
+
+    /**
+     * The read limit
+     */
+    protected int markLimit;
+
+    /**
+     * The position
+     */
+    protected int pos;
+
+    private static final int BUFFER_SIZE = 8192;
+
+    private final IndexRoutingTableHeader indexRoutingTableHeader;
+
+    private final Iterator<IndexShardRoutingTable> shardIter;
+
+    public IndexRoutingTableInputStream(IndexRoutingTable indexRoutingTable, long version, Version nodeVersion) throws IOException {
+        this(indexRoutingTable, version, nodeVersion, BUFFER_SIZE);
+    }
+
+    public IndexRoutingTableInputStream(IndexRoutingTable indexRoutingTable, long version, Version nodeVersion, int size)
+        throws IOException {
+        this.buf = new byte[size];
+        this.shardIter = indexRoutingTable.iterator();
+        this.indexRoutingTableHeader = new IndexRoutingTableHeader(version, indexRoutingTable.getIndex().getName(), nodeVersion);
+        initialFill();
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (pos >= count) {
+            maybeResizeAndFill();
+            if (pos >= count) return -1;
+        }
+        return buf[pos++] & 0xff;
+    }
+
+    private void initialFill() throws IOException {
+        BytesReference bytesReference = indexRoutingTableHeader.write();
+        buf = bytesReference.toBytesRef().bytes;
+        count = bytesReference.length();
```

Comment:
With this the initialization of `buf` in ctor is not used. Also, the buffer size will not be `size` from this point onwards but would be `count`